samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,340 @@
  */
 package org.apache.samza.operators.impl;
 
-import org.apache.samza.config.Config;
-import org.apache.samza.operators.MessageStreamImpl;
 import org.apache.samza.operators.spec.WindowOperatorSpec;
+import org.apache.samza.operators.WindowState;
+import org.apache.samza.operators.triggers.Cancellable;
+import org.apache.samza.operators.triggers.RepeatingTriggerImpl;
+import org.apache.samza.operators.triggers.TimeTrigger;
+import org.apache.samza.operators.triggers.Trigger;
+import org.apache.samza.operators.triggers.TriggerContext;
+import org.apache.samza.operators.triggers.TriggerImpl;
+import org.apache.samza.operators.triggers.TriggerImpls;
+import org.apache.samza.operators.triggers.FiringType;
+import org.apache.samza.operators.util.InternalInMemoryStore;
+import org.apache.samza.operators.windows.AccumulationMode;
+import org.apache.samza.operators.windows.WindowKey;
 import org.apache.samza.operators.windows.WindowPane;
 import org.apache.samza.operators.windows.internal.WindowInternal;
+import org.apache.samza.operators.windows.internal.WindowType;
+import org.apache.samza.storage.kv.KeyValueStore;
 import org.apache.samza.task.MessageCollector;
-import org.apache.samza.task.TaskContext;
 import org.apache.samza.task.TaskCoordinator;
+import org.apache.samza.util.Clock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.Function;
+
+/**
+ * Implementation of a window operator that groups messages into finite windows for processing.
+ *
+ * This class implements the processing logic for various types of windows and triggers. It tracks and manages state for
+ * all open windows, the active triggers that correspond to each of the windows and the pending callbacks. It provides
+ * an implementation of {@link TriggerContext} that {@link TriggerImpl}s can use to schedule and cancel callbacks. It
+ * also orchestrates the flow of messages through the various {@link TriggerImpl}s.
+ *
+ * <p> An instance of a {@link TriggerImplWrapper} is created corresponding to each {@link Trigger} configured for a
+ * particular window. For every message added to the window, this class looks up the corresponding {@link TriggerImplWrapper}
+ * for the trigger and invokes {@link TriggerImplWrapper#onNext(Object, MessageCollector, TaskCoordinator)}. The
+ * {@link TriggerImplWrapper} forwards the callback to its {@link TriggerImpl} instance. A {@link TriggerImpl} instance is scoped
```

Comment:
I thought more about this after our discussion. `triggerImplState.onTimer()` and `triggerImplState.onMessage()` felt awkward. All that is doing, is wrapping the `TriggerImpl` and forwarding the callback to its underlying `TriggerImpl`.